agenthub/monologue_agent/utils/memory.py

```diff
@@ -128,11 +133,14 @@ def add_event(self, event: dict):
         )
         self.thought_idx += 1
         logger.debug('Adding %s event to memory: %d', t, self.thought_idx)
-        thread = Thread(target=self._add_doc, args=(doc,))
+        thread = threading.Thread(target=self._add_doc, args=(doc,))
+        self._add_threads.append(thread)
         thread.start()  # We add the doc concurrently so we don't have to wait ~500ms for the insert
 
     def _add_doc(self, doc):
+        sema.acquire()
         self.index.insert(doc)
+        sema.release()
```

Comment:
```suggestion
        with sema:
            self.index.insert(doc)
```

If `self.index.insert(doc)` throws an error, then `sema.release()` won't be called. Wrapping in `try/finally` works but IMO a context manager is the most concise.